dev/plugins/hu.elte.txtuml.validation/src/hu/elte/txtuml/validation/sequencediagram/visitors/Utils.java

```diff
@@ -0,0 +1,169 @@
+			if (modifier instanceof Annotation) {
+				Annotation ca = (Annotation) modifier;
+				if (ca.getTypeName().getFullyQualifiedName().equals("Position")) {
+					// position value
+					SingleMemberAnnotation position;
+					try {
+						position = (SingleMemberAnnotation) ca;
+					} catch (ClassCastException ex) {
+						return;
+					}
+					int annotationVal = (int) position.getValue().resolveConstantExpressionValue();
+					if (annotationVal < 0) {
+						collector.report(SequenceErrors.INVALID_POSITION.create(collector.getSourceInfo(), position));
+					}
+
+					// type of annotated field
+					boolean isModelClass = ElementTypeTeller.hasSuperClass(elem.getType().resolveBinding(),
+							ModelClass.class.getCanonicalName());
+					if (!isModelClass) {
+						collector.report(SequenceErrors.INVALID_LIFELINE_DECLARATION.create(collector.getSourceInfo(),
+								elem.getType()));
+					}
+				}
+			}
+		}
+	}
+
+	public static void checkSendExists(ProblemCollector collector, Block node, ASTVisitor visitor) {
+		List<Statement> statements = (List<Statement>) node.statements();
+		boolean isLeaf = !statements.stream().anyMatch(stm -> stm instanceof WhileStatement
+				|| stm instanceof IfStatement || stm instanceof ForStatement || stm instanceof DoStatement);
+		if (!isLeaf) {
+			return;
+		}
+
+		List<MethodInvocation> methodInvocations = statements.stream().map(stm -> ((Statement) stm))
+				.filter(stm -> stm instanceof ExpressionStatement).map(stm -> (ExpressionStatement) stm)
+				.map(ExpressionStatement::getExpression).filter(expr -> expr instanceof MethodInvocation)
+				.map(expr -> (MethodInvocation) expr).collect(toList());
+
+		List<MethodInvocation> notSequenceInvocations = methodInvocations.stream().filter(inv -> {
+			try {
+				String declaringClass = inv.resolveMethodBinding().getDeclaringClass().getQualifiedName();
+				return !declaringClass.equals(Sequence.class.getCanonicalName())
+						&& !declaringClass.equals(Action.class.getCanonicalName());
+			} catch (NullPointerException ex) {
+				return false;
+			}
+		}).collect(toList());
+
+		boolean containsSend = methodInvocations.stream().anyMatch(Utils::isSendInvocation);
+		if (!containsSend) {
+			boolean isAllSequenceMethod = methodInvocations.stream().allMatch(inv -> {
+				try {
+					return inv.resolveMethodBinding().getDeclaringClass().getQualifiedName()
+							.equals(Sequence.class.getCanonicalName());
+				} catch (NullPointerException ex) {
+					return false;
+				}
+			});
+
+			if (methodInvocations.isEmpty() || isAllSequenceMethod) {
+				collector.report(SequenceErrors.SEND_EXPECTED.create(collector.getSourceInfo(), node));
+			}
+			notSequenceInvocations.forEach(inv -> validateInvokedMethod(inv, visitor, collector));
+		}
+	}
+
+	public static void checkInvalidActionCall(MethodInvocation elem, ProblemCollector collector) {
+		boolean isActionCall;
+		try {
+			isActionCall = elem.resolveMethodBinding().getDeclaringClass().getQualifiedName()
```

Comment:
This catches only direct action calls, i.e. invoking `void f() { Action.log(""); }` is considered valid but it's clearly not.

Reply:
Dealing with this issue is not trivial at all. I think this solution would do for now as this case is handled with an exception during runtime. Finding a better solution is going to be a future work.

Reply:
I agree.